### ISSUE TITLE
When using `relationMode="prisma"`or for MongoDB, provide an escape hatch to remove emulated checks from `onUpdate` and `onDelete`

### ISSUE THREAD

**Issue by Jolg42** [prisma/prisma#15759]
## Problem

We have a preview feature called `referentialIntegrity`
It allows users to set the datasource property `relationMode` to `"prisma"` or `"foreignKeys"`
When `relationMode="prisma"` 
	- Prisma Client emulates the behavior of the referential actions, `onUpdate`, `onDelete` based on the behavior observed in SQL databases that use Foreign Keys.
	- Prisma Migrate does not create Foreign Keys,
For MongoDB provider, `relationMode="prisma"` is the default and cannot be changed, it makes it easier to work with relations.

In some cases, people maybe want to have an alternative to this emulated behavior of Prisma Client
- In MongoDB world this is a new behavior
	- some people might prefer handling relations on their own
		- which I think make sense if the MongoDB database is accessed by another ORM/Client (which does not have the same behavior).
	- some people might find it slower because of additional queries 

## Option 1 `relationMode="none"`
- Valid for all databases
- It would be a global setting where all referential actions would be invalid and show a schema validation error.
- The behavior would be that the default `onUpdate`, `onDelete` does "nothing", no checks would be performed, which means that you could break the Referential Integrity of your data. (The referenced entity might not exist)

## Option 2 `onUpdate="Nothing" / onDelete="Nothing"`
- Valid for all databases
- Only valid if `relationMode="prisma"`
- It would be a Prisma "referential action" that only exist for Prisma
- The behavior would be that `onUpdate`, `onDelete` does "nothing", no checks would be performed, which means that you could break the Referential Integrity of your data. (The referenced entity might not exist)

**Comment by jkomyno** [comment]
Note, as per what @Jolg42 said yesterday on Zoom, this use case would not only be of interest for `mongodb` users, but for SQL database users as well. In particular, there are users who are unaware of the semantic implications of `NoAction`, and confuse it for "a referential action that does nothing".

**Comment by janpio** [comment]
Is the desired behavior possible with relational databases when using `relationMode="foreignKeys"` (default, where relations are managed via foreign keys)?

**Comment by Jolg42** [comment]
Removing emulated checks brought by `relationMode = "prisma"` is indeed possible on relational databases, only if you do not use `@relation` but then you lose the Prisma Client API that comes with it.

**Comment by Jolg42** [comment]
Notes: 
- For MongoDB what people might want to disable is the emulation on `update` and `delete` Prisma Client queries, (create is not covered by the emulation).
Technically, we can do that by introducing `Nothing`, for `onUpdate` `onDelete`, on `@relation`, where both should be set?

- When `relationMode="foreignKeys"` (default), on relational databases.
`Nothing` would be impossible (as long as those are real relations with foreign keys).
Which makes us think that something like a "weak" relation could be an option https://github.com/prisma/prisma/issues/7351

**Comment by davesteinberg** [comment]
> * When `relationMode="foreignKeys"` (default), on relational databases.
>   `Nothing` would be impossible (as long as those are real relations with foreign keys).
>   Which makes us think that something like a "weak" relation could be an option [Support (some) weak relations with no foreign key #7351](https://github.com/prisma/prisma/issues/7351)

Just a note that using foreign keys in general, with the ability to opt out for some particular relations (aka "weak" relations) is still important for me. I'm still removing my weak relations for `prisma migrate` and adding them back for `prisma generate`.

So, it would be super helpful to have the migrate-time escape hatch for foreign keys, as well.

**Comment by davesteinberg** [comment]
@janpio Yup, I just meant to reaffirm the need and agree that I see the connection to this. Sorry for the noise.

**Comment by Songkeys** [comment]
I’m unfortunately blocked from upgrading prisma@4.4.0 to a newer version. Because I need to reply on the “bug-as-a-feature” behaviour on v4.4.0 that “NoAction” actually does nothing, while "NoAction" becomes "Restrict" since v4.5.0. So I really need this "Nothing" feature.

**Comment by janpio** [comment]
Are you ok with your data getting invalid and you having orphan entries after? What is your motivation not to prevent such actions or removing the invalid valid after? (Just trying to understand)

**Comment by Songkeys** [comment]
> Are you ok with your data getting invalid and you having orphan entries after? What is your motivation not to prevent such actions or removing the invalid valid after? (Just trying to understand)

For example, I have `user` and `post` tables. A user will post many posts. But there might be some case where a post is coming to the database while its owner hasn't been inserted into. Because the data was scrape from the internet, blockchain, random other places. I have no control or idea what data would come or lose first actually. So when inserting, updating, deleting, I need to only update the row itself, not to influence other tables. But when query, they do have relations, which needs the usage of relationMode="prisma".

I think this scene is pretty common? Orphan entries are totally acceptable. ORM doesn't need to assume that we must update or delete the related table's row if one row is changed.

**Comment by Songkeys** [comment]
Hi Team. May I know the status of this issue? This prevents me updating from 4.4.0 to enjoy the latest features ever (will get error 2014 "The change you are trying to make would violate the required relation 'ChildToParent' between the `Child` and `Parent` models."). Thank you.

**Comment by janpio** [comment]
It is not on the top of our list and hence not prioritized. We can not say if or when this will be implemented.

**Comment by kibertoad** [comment]
> Are you ok with your data getting invalid and you having orphan entries after? What is your motivation not to prevent such actions or removing the invalid valid after? (Just trying to understand)

Yes, in many cases it's perfectly fine to rely on your transactions to be implemented correctly and application ensuring the referential integrity. In heavy load systems it is sometimes crucial to turn off any strong integrity constraints in order to scale.

Currently the only escape hatch is raw queries, and that becomes cumbersome for complex relationships very quickly.

**Comment by ItzSiL3Nce** [comment]
This issue would _probably_ be closed by https://github.com/prisma/prisma-engines/pull/4116
